### ISSUE TITLE
`rw` duplicates goals for metavariables appearing in the rewritten expression

### ISSUE THREAD

**Issue by JovanGerb** [leanprover/lean4#4381]
### Prerequisites

Please put an X between the brackets as you perform the following steps:

* [x] Check that your issue is not already filed:
      https://github.com/leanprover/lean4/issues
* [x] Reduce the issue to a minimal, self-contained, reproducible test case.
      Avoid dependencies to Mathlib or Batteries.
* [x] Test your test case against the latest nightly release, for example on
      https://live.lean-lang.org/#project=lean-nightly
      (You can also use the settings there to switch to “Lean nightly”)

### Description

When rewriting a term that contains metavariables, this metavariable is added as a new goal in the tactic state, even if it already is a goal. This clutters the tactic state and is confusing.

### Context

Question on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Rewriting.20the.20goal.20with.20metavariables.20duplicates.20goals/near/442515619)

### Steps to Reproduce

The example provided on Zulip is:
```
example : ∀ d g, d = g → exists x : Nat, x = d := by
  intros d g H1
  constructor
  rewrite [H1,←H1,H1,←H1,H1]
  -- rotate_left; exact g; rfl
```

**Expected behavior:** [Clear and concise description of what you expect to happen]

The goal
```
dg: Nat
H1: d = g
⊢ Nat
```
should still appear only once after the rewrites

**Actual behavior:** [Clear and concise description of what actually happens]

The goal appears 6 times after the 5 rewrites.

### Versions

4.7.0 (works on any version)

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by JovanGerb** [comment]
I've made a fix for it here: #4383

**Comment by eric-wieser** [comment]
Either this has regressed, or the fix wasn't exhaustive: see #6407
